Fix ssh-agent use by libgit2

Keno opened the pull request:
cc @wildart. Works for me, but I'm not convinced this is the right solution.

tkelman commented on base/libgit2/callbacks.jl:
```diff
@@ -82,7 +82,8 @@ function credentials_callback(cred::Ptr{Ptr{Void}}, url_ptr::Cstring,
 
         # first try ssh-agent if credentials support its usage
-        if creds[:usesshagent, credid] == "Y"
+        if creds[:usesshagent, credid] === nothing ||
```
shouldn't this have just happened a few lines up?

Keno replied:
That's what I thought, but if creds is EmpyCredentials, it ignores that setindex.

tkelman replied:
that sounds like a bug maybe?

Keno replied:
Well, it looks to be by design, but I'm not sure, @wildart should take a look.

wildart replied:
no need, https://github.com/JuliaLang/julia/blob/c351ed6071a909a9dba70fc127c4bc40671bbaa3/base/libgit2/callbacks.jl#L80

wildart replied:
Right, `EmpyCredentials` is a useless class. But this fix will work.

tkelman replied:
why is it getting used or passed around then? if it doesn't store anything, then all the logic is going to be wrong for it amd this is not going to be the only bug

Keno replied:
Yeah, having a setindex that is later not retrievable by setindex with the same index violates expectations. It would be good to redesign that API

tkelman replied:
right. don't need to do so right this minute unless that turns out to be the best fix to the blocker bugs we're seeing, but it's going to be error prone until we do

wildart commented:
You better disable ssh-agent usage on windows.

tkelman commented:
how should windows users use ssh without it? cc @davidanthoff

wildart commented:
Not all git builds on windows come with ssh-agent. I believe recent windows build of Git integrates with  Windows Credential Store. I'm not sure whether it is used for ssh.

Keno commented:
I don't understand what the harm is. If the agent can't be found, it should fall back to alternative methods.

davidanthoff commented:
Here is my understanding of the landscape. I might well be off. Current git versions for windows bring ssh-agent along, i.e. ssh-agent.exe is in the usr\bin folder in the folder where you install git. posh-git (the powershell plugin that most people seem to use if the use powershell and git), automatically starts ssh-agent when you start a new powershell prompt. So when I start julia from a powershell prompt, ssh-agent is running. Once ssh-agent is started, you have to manually kill it in task manager, or log out, to end that process. So at least for me, ssh-agent is running 99% of the time.

If I'm in a window session and had never started powershell (with posh-git) and had not somehow manually started ssh-agent, it would probably not be running. If I then started julia from the startmenu shortcut, I guess ssh-agent would not be running.

[This](https://github.com/git-for-windows/git/wiki/Auto-launching-ssh-agent-when-git-starts) has more info that might be relevant. And [this](https://github.com/git-for-windows/git/wiki/OpenSSH-Integration-with-Pageant). Then there is also the [Git credential manager for Windows](https://github.com/Microsoft/Git-Credential-Manager-for-Windows).

So, my sense is that a lot of Windows users will have a ssh-agent running. I think if it is running, julia should just use it, and not prompt for anything, right?

tkelman commented:
if you check ssh-agent.exe in dependency walker though, it should be linking to msys-2.0.dll, or cygwin1.dll for cygwin's ssh-agent, so it's a posix executable rather than a win32 native one. we no longer bundle a posix layer with the windows julia install (now that cygwin is lgpl we could consider it), so any ssh-agent (or pageant) would have to be a bring your own deal. very few windows users know anything about powershell based tools in my experience, so you might be overestimating how common this will be.

davidanthoff commented:
I think the following would be perfectly fine: if julia sees an ssh-agent running, use it, if not, don't. That is probably how it works on Linux and Mac already, right? On Windows that would essentially mean that those people that have git installed (which ships ssh-agent) *and* got the agent started somehow would see julia using it. Git itself seems the most prevalent vehicle for users to get the ssh-agent binaries on their machines, and then there are lots of ways on how it might be started: the github help tells you how to start it, posh-git starts it, if you install git via something like the github desktop (or whatever the name is), it will run automatically etc.

tkelman commented:
I need to check the code again for how it actually uses ssh-agent, but there's a difference between checking if an ssh-agent is running vs if the ssh-agent executable is on the path. It almost always will be on the path on osx and linux, and often won't be on windows. I think most pieces of software should be much more cautious about modifying users' path settings.

wildart commented:
Libgit2 gives very little leeway in terms of managing user credentials. So, either we provide some sort of setup procedure for using SSH protocol, or presume some default parameters (e.g. location of the key), or let user set it up through environmental variables. One should not expect same level of UX functionality from `libgit2` comparing to `git`.

Keno commented:
@wildart Do you see this as the correct fix or is there something else we should do?

wildart commented:
So, I wanted to connect to ssh-agent firstly before falling back on manually specified ssh key. This PR would always request ssh-agent authentication in `usesshaagent` field is not provided in `Credentials` object, and would not provide any alternative authentication mechanism.

[Failing to connect to ssh-agent](https://github.com/JuliaLang/julia/pull/17391#issuecomment-233151415) should not be treaded as error. It only means that some setup is required before using SSH protocol in the package manager.

Keno commented:
Yes, if the ssh-agent doesn't work, it should fall back to something else. If this is not the correct fix, please suggest an appropriate one.

wildart commented:
There is one caveat: If ssh-agent authentication fails with error, whole operation is aborted.

Keno commented:
What counts as an error?

wildart commented:
https://github.com/JuliaLang/julia/pull/17391#issuecomment-233151415

Keno commented:
That's pretty stupid. Can we patch libgit2 to avoid this?

wildart commented:
Sure, given some time, we can do anything. Just for ref: https://github.com/libgit2/libgit2/issues/3761

Keno commented:
It just seems ridiculous to me that there's no way to retry a request if the ssh agent can't be found.

wildart commented:
A bit of topic, I looked at `libssh2` code and found that this error https://github.com/JuliaLang/julia/pull/17391#issuecomment-233151415 related to `Pageant`. It looks like ssh-agent is not used for windows, `pageant` on other hand is.

wildart commented:
The proper solution would be to return a authentication error to callback and not fail until callback will specify so.

Keno commented:
Maybe re-open the issue with an explanation of the use case we have and why the alternate behavior would be desirable?

wildart commented:
I reopend https://github.com/libgit2/libgit2/issues/3761.

Keno commented:
Let's just temporarily patch [this loop](https://github.com/libgit2/libgit2/blob/master/src/transports/ssh.c#L611) to retry if there was an error with the agent.

Keno committed:
Remove EmptyCredentials and fix ssh-agent during Pkg.clone

Also make setindex! fail if requested on a credential type that does
not have the given field.

tkelman commented on base/libgit2/types.jl:
```diff
@@ -56,19 +56,16 @@ abstract AbstractPayload
-    end
+function Base.setindex!(p::AbstractCredentials, val, key, _=nothing)
+    ks = Symbol(key)
+    @assert isdefined(p, ks)
```
better error here?

Keno replied:
It's an internal error, an assert is fine.

tkelman replied:
if it ever happens, would want to be able to see what the inputs were

Keno commented:
I've pushed an update to this, removing EmptyCredentials entirely. I'll test on windows to see what we can do about failure to connect to ssh agent.

Keno committed:
Make failure to connect to ssh-agent non-fatal

tkelman commented on deps/libgit2.mk:
```diff
@@ -7,7 +7,7 @@ $(eval $(call git-external,libgit2,LIBGIT2,CMakeLists.txt,build/libgit2.$(SHLIB_
 LIBGIT2_OBJ_TARGET := $(build_shlibdir)/libgit2.$(SHLIB_EXT)
 
-LIBGIT2_OPTS := $(CMAKE_COMMON) -DCMAKE_BUILD_TYPE=Release -DTHREADSAFE=ON
+LIBGIT2_OPTS := $(CMAKE_COMMON) -DCMAKE_BUILD_TYPE=Debug -DTHREADSAFE=ON
```
temporary?

Keno replied:
Yes, left over from debugging, sorry.

Keno commented:
Ok, now with 100% fewer fatal failures due to ssh-agent.

tkelman commented:
@Keno did this patch ever get upstreamed? If not, any reason it didn't?

Keno commented:
IIRC there was an issue opened (https://github.com/libgit2/libgit2/issues/3866), but this was more of a workaround than a real fix.

tkelman commented:
Oh right, thanks, I even bumped it a few months later. Maybe I'll ping them on their slack to see if I can get an opinion.
